keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,6 +333,113 @@ myOtherField: newly added field
 The feature proposed in this KEP would allow this change, since `myField` was
 unaltered from the original object. If `myField` was changed, the new value would have to pass the new validation rule to be written.
 
+### Nested Value Validations 
+
+OpenAPI-schemas support nested validations with logical operations:
+- not
+- oneOf
+- anyOf
+- allOf
+
+To keep the semantics simple and implementation clear, schemas and validations
+nested within the following will be not ratcheted:
+
+- not
+- oneOf
+
+Our intuition is that in the wild CRD authors do not make much use of them, and
+when they do the logic is complicated and references multiple fields. 
+
+For all these constructs, the conservative `DeepEqual` rule still applies 
+to the final output of the rule. 
+
+The following will be ratcheted with the conservative ratcheting rule, and
+their children will also be ratcheted.
+
+- allOf
+- anyOf
+
+### Atomic Lists and Maps
+
+Following Kubernets convention that fields annotated `x-kubernetes-list-type` or `x-kubernetes-map-type` as `atomic` change as a unit, the ratcheting rules
+do not apply to values nested inside atomic lists and maps. But a DeepEqual 
+comparison on the entire object can still ratchet errors for subfields.
+
+### Weaker Ratcheting Rule for Value Validations
+
+Above we propose a conservative catch-all ratcheting rule that should provide
+a minimum level of ratcheting support for any validation rule in the schema.
+
+For most types of validations, the `DeepEqual(OldObject, NewObject)` rule is
+strong enough, but for others it is too weak. Consider the `required` error. This
+validation is applied to the object owning the property. Thus, if a new field
+becomes required the object cannot have other fields added onto it without also 
+adding the required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+The following fields are all the value validations allowed in our schemas:
+
+- Format
+- Maximum
+- ExclusiveMaximum
+- Minimum
+- ExclusiveMinimum
+- MaxLength
+- MinLength
+- Pattern
+- MultipleOf
+- Enum
+- Nullable
```

Comment:
not sure it matters, but Nullable is technically not a value validation in Kube. E.g. you cannot write `not nullable`.

Reply:
moved nullable to separate list from the value validations